rlog/2024-12-23-merkle-tree.mdx

```diff
@@ -0,0 +1,213 @@
+---
+title: 'Vac 101 - Climbing Merkle Trees'
+date: 2024-12-23 12:00:00
+authors: marvin
+published: true
+slug: climbing-merkle-trees
+categories: research
+
+toc_min_heading_level: 2
+toc_max_heading_level: 4
+---
+
+In this post, we introduce a crucial data structure used throughout web3.
+
+<!--truncate-->
+
+## Introduction
+
+A large amount of data is swapped between users on a blockchain in the form of transactions.
+Over the entire life of a blockchain,
+the storage space required to maintain a copy of every transaction becomes untenable for most users.
+However, the integrity of a blockchain relies on a large pool of users that can validate the blockchain's history from its inception to its present state.
+The data representing the blockchain's state is compressed.
+This compression addresses the issue of scalability that would otherwise greatly restrict the pool of users.
+
+Data compression alone is not the end goal.
+As mentioned, it is essential for users to be able to validate the blockchain's history.
+The property of compression and validation was solved in Bitcoin by the use of Merkle trees.
+Merkle trees were first by Ralph Merkle in his dissertation [[1](https://www.ralphmerkle.com/papers/Thesis1979.pdf)].
+A Merkle tree is a data structure that compresses a digest of data to a constant size while still providing a method for proving membership of elements of the digest.
+A previous rlog[[2](https://vac.dev/rlog/rln-light-verifiers/)] described how Merkle trees, with their proof of membership, could be used for lightweight clients for RLN. 
+
+## Tree structure
```

Comment:
CharGPT advises the following definition of trees: "A tree is a special data structure that organizes nodes in a hierarchical manner such that there is exactly one path between any two nodes, and there are no cycles."

Also it gives two important remarks: 

"The statement about leaves is somewhat misleading. Leaves are indeed nodes that do not have any children, but they are not necessarily the highest level of the tree. Instead, they are typically found at the lowest level (or levels) of the tree, as they are terminal nodes.

"The term "parentless nodes" might confuse readers. In tree terminology, leaves are nodes that do not have children, while the root node is the only node without a parent."

Thus, fixing the terminology and mistype regarding childless nodes is required.

Reply:
pt1. ChatGPT's definition of tree. "Exactly one path between two nodes" is equivalent to no cycles. Intuitively, I think a path is more clear to nontechnical readers than a cycle.

pt2. Often, Merkle trees are often presented as an inverted Merkle tree. For Merkle trees (with 2^n leaves) it is customary to refer to the root node in layer 0 and leaves in layer n. This is usually done so that a binary expansion of the leaf's index can be used (in a natural manner) to dictate whether the path.

pt3. Fixed the parentless node to childless. This was a carry over to an early draft based on how Merkle trees are usually presented (see pt2).

Reply:
1. Agree. To be more precise, I would clarify that the trees we consider here are rooted trees and add their definition "A rooted tree is a tree in which one vertex has been designated the root."

2. I guess it is ok, but then we should consider Merkle tree to be growing in depth, i.e. the lower is some layer the larger its number.

3. Good.